Add ability to flag classes as experimental or deprecated

SaracenOne opened the pull request:
Draft WIP PR. Only supports is_experimental and is_deprecated flags for whole classes at the moment, but hoping to add support for methods, properties, and signals soon.

Implementing the proposal described here [#5279](https://github.com/godotengine/godot-proposals/issues/5279)

SaracenOne committed:
Add ability to flag classes as experimental or deprecated.

Calinou commented on editor/create_dialog.cpp:
```diff
@@ -300,6 +300,15 @@ void CreateDialog::_configure_search_option_item(TreeItem *r_item, const String
+	bool is_experimental = EditorHelp::get_doc_data()->class_list[p_type].is_experimental;
+
+	if (is_deprecated) {
+		r_item->add_button(0, get_theme_icon("NodeWarning", SNAME("EditorIcons")), 0, false, TTR("This class is marked as deprecated."));
```
I'd use the `StatusError` icon for deprecated classes to distinguish the warning from the experimental one.

Calinou replied:
I'd use the `StatusError` icon for deprecated classes to distinguish the warning from the experimental one.

```suggestion
		r_item->add_button(0, get_theme_icon("StatusError", SNAME("EditorIcons")), 0, false, TTR("This class is marked as deprecated."));
```

Calinou commented on editor/editor_help.cpp:
```diff
@@ -610,6 +616,26 @@ void EditorHelp::_update_doc() {
 
+	// Note if deprecated.
+	if (cd.is_deprecated) {
+		class_desc->push_color(get_theme_color(SNAME("warning_color"), SNAME("Editor")));
```
Use error_color to match the above icon change suggestion:

```suggestion
		class_desc->push_color(get_theme_color(SNAME("error_color"), SNAME("Editor")));
```

Calinou commented on editor/editor_help.cpp:
```diff
@@ -553,6 +553,12 @@ void EditorHelp::_update_doc() {
 	class_desc->pop(); // font
+
+	Ref<Texture2D> warning_icon;
+	if (cd.is_deprecated || cd.is_experimental) {
+		warning_icon = get_theme_icon(SNAME("NodeWarning"), SNAME("EditorIcons"));
+		class_desc->add_image(warning_icon, warning_icon->get_width(), warning_icon->get_height());
+	}
```
Use the StatusError icon for `cd.is_deprecated` only as mentioned above.

Calinou commented on editor/editor_help_search.cpp:
```diff
@@ -570,6 +570,15 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
 	}
 
+	if (p_doc->is_deprecated || p_doc->is_experimental) {
+		Ref<Texture2D> warning_icon = ui_service->get_theme_icon("NodeWarning", SNAME("EditorIcons"));
```
Use the StatusError icon for `cd.is_deprecated` only as mentioned above.

Calinou commented on scene/main/node.h:
```diff
@@ -531,4 +531,8 @@ Error Node::rpc_id(int p_peer_id, const StringName &p_method, VarArgs... p_args)
 }
 
+// Add these macro to your class's 'get_configuration_warnings' function to have warnings show up in the scene tree inspector.
+#define DEPRECATED_NODE_WARNING warnings.push_back(RTR("This node is marked as deprecated and will be removed in future versions. Please check the Godot documentation for information how to migrate."));
```
Add a line break to prevent the node configuration warning from being too long:

```suggestion
#define DEPRECATED_NODE_WARNING warnings.push_back(RTR("This node is marked as deprecated and will be removed in future versions.\nPlease check the Godot documentation for information how to migrate."));
```

SaracenOne commented:
Thanks for the feedback!

SaracenOne commented:
Latest update: 
- Implemented Calinou's recommendations for deprecation.
- Added support to class members too.
- Removed deprecation and experimental flags from ClassDB since I think it can be implemented well enough purely as a doc feature.

SaracenOne commented:
I think this is at a point it can be considered merging now, though it might be worth asking what things, if any, might be appropriate candidates for this flag. Given 4.0 is a clean slate of sorts, deprecation isn't concern, but there are certain things which might be worth considering for 'experimental' status given the uncertainity surrounding their implementation.

akien-mga commented on editor/doc_tools.cpp:
```diff
@@ -1274,6 +1317,14 @@ Error DocTools::_load(Ref<XMLParser> parser) {
 								if (parser->has_attribute("is_bitfield")) {
 									constant2.is_bitfield = parser->get_attribute_value("is_bitfield").to_lower() == "true";
 								}
+								if (parser->has_attribute("is_deprecated")) {
+									String result = parser->get_attribute_value("is_deprecated");
+									constant2.is_deprecated = (result == "true" || result == "True" || result == "TRUE" || result == "1");
+								}
+								if (parser->has_attribute("is_experimental")) {
+									String result = parser->get_attribute_value("is_experimental");
+									constant2.is_experimental = (result == "true" || result == "True" || result == "TRUE" || result == "1");
+								}
```
You could use `parser->get_attribute_value("is_experimental").to_lower() == "true"` like above.

The XML docs are not meant to be edited by hand so in theory we only need to support what we write. We don't have to support `1`.

akien-mga commented on editor/editor_help.cpp:
```diff
@@ -335,6 +352,14 @@ void EditorHelp::_add_method(const DocData::MethodDoc &p_method, bool p_overview
 		class_desc->pop();
 	}
 
+	if (p_method.is_deprecated) {
+		DEPRECATED_DOC_TAG
+	}
+
+	if (p_method.is_experimental) {
+		EXPERIMENTAL_DOC_TAG
+	}
```
It's not needed technically but macros look cleaner if they look like actual statements IMO.
```suggestion
	if (p_method.is_deprecated) {
		DEPRECATED_DOC_TAG;
	}

	if (p_method.is_experimental) {
		EXPERIMENTAL_DOC_TAG;
	}
```

akien-mga reviewed (approved):
Looks good to me overall.

I'm not convinced that we need the level of detail added here where properties, methods, signals and constants can be marked as experimental. But I guess we can always reconsider in the future and see if we did need this granularity or if we're happy just flagging classes only.

SaracenOne commented:
Rebased now with changes recommended by @akien-mga.

SaracenOne commented:
Though, actually, I'm not sure if I'm specifically being asked to remove the deprecation/experimental markings for members/methods/signals. I personally don't see harm in having the markings avalible for the sake of consistency and the likelyhood that we may be more inclined deprecate individual functions as opposed to whole classes, but I guess can strip this part if you'd like.

akien-mga commented:
Thanks!
